aten/src/ATen/cpu/vec256/vec256_base.h

```diff
@@ -64,115 +66,199 @@ struct Vec256 {
+}
+
+template <class T>
+Vec256<T> ceil(Vec256<T> x) {
+  return map(std::ceil, x);
+}
+
+template <class T>
+Vec256<T> cos(Vec256<T> x) {
+  return map(std::cos, x);
+}
+
+template <class T>
+Vec256<T> floor(Vec256<T> x) {
+  return map(std::floor, x);
+}
+
+template <class T>
+Vec256<T> round(Vec256<T> x) {
+  return map(std::round, x);
+}
+
+template <class T>
+Vec256<T> sin(Vec256<T> x) {
+  return map(std::sin, x);
+}
+
+template <class T>
+Vec256<T> sqrt(Vec256<T> x) {
+  return map(std::sqrt, x);
+}
+
+template <class T>
+Vec256<T> neg(Vec256<T> x) {
+  return Vec256<T>(0) - x;
+}
+
+template <class T>
+Vec256<T> reciprocal(Vec256<T> x) {
+  Vec256<T> ret;
+  for (int64_t i = 0; i < Vec256<T>::size; i++) {
+    ret.set_value(i, 1 / x[i]);
   }
-  Vec256<T> sqrt() const {
-    return map(std::sqrt);
+  return ret;
+}
+
+template <class T>
+Vec256<T> rsqrt(Vec256<T> x) {
+  return reciprocal(sqrt(x));
+}
+
+template <class T>
+Vec256<T> sigmoid(Vec256<T> x) {
+  Vec256<T> ret;
+  for (int64_t i = 0; i < Vec256<T>::size; i++) {
+    ret.set_value(i, ((T)1.0) / (((T)1.0) + std::exp(-x[i])));
   }
-};
+  return ret;
+}
+
+template <class T>
+Vec256<T> tanh(Vec256<T> x) {
+  return map(std::tanh, x);
+}
+
+template <class T>
+Vec256<T> trunc(Vec256<T> x) {
+  return map(std::trunc, x);
+}
+
+template <class T>
+Vec256<T> frac(Vec256<T> x) {
+  return x - trunc(x);
+}
 
 template <class T> Vec256<T> operator+(const Vec256<T> &a, const Vec256<T> &b) {
-  Vec256<T> c = Vec256<T>();
+  Vec256<T> c;
   for (int i = 0; i != Vec256<T>::size; i++) {
-    c.values[i] = a.values[i] + b.values[i];
+    c.set_value(i, a[i] + b[i]);
   }
   return c;
 }
 
 template <class T> Vec256<T> operator-(const Vec256<T> &a, const Vec256<T> &b) {
-  Vec256<T> c = Vec256<T>();
+  Vec256<T> c;
   for (int i = 0; i != Vec256<T>::size; i++) {
-    c.values[i] = a.values[i] - b.values[i];
+    c.set_value(i, a[i] - b[i]);
   }
   return c;
 }
 
 template <class T> Vec256<T> operator*(const Vec256<T> &a, const Vec256<T> &b) {
-  Vec256<T> c = Vec256<T>();
+  Vec256<T> c;
   for (int i = 0; i != Vec256<T>::size; i++) {
-    c.values[i] = a.values[i] * b.values[i];
+    c.set_value(i, a[i] * b[i]);
   }
   return c;
 }
 
 template <class T> Vec256<T> operator/(const Vec256<T> &a, const Vec256<T> &b) {
-  Vec256<T> c = Vec256<T>();
+  Vec256<T> c;
   for (int i = 0; i != Vec256<T>::size; i++) {
-    c.values[i] = a.values[i] / b.values[i];
+    c.set_value(i, a[i] / b[i]);
   }
   return c;
 }
 
 template <class T> Vec256<T> max(const Vec256<T> &a, const Vec256<T> &b) {
-  Vec256<T> c = Vec256<T>();
+  Vec256<T> c;
+  for (int i = 0; i != Vec256<T>::size; i++) {
+    c.set_value(i, std::max(a[i], b[i]));
+  }
+  return c;
+}
+
+template <class T> Vec256<T> min(const Vec256<T> &a, const Vec256<T> &b) {
+  Vec256<T> c;
   for (int i = 0; i != Vec256<T>::size; i++) {
-    c.values[i] = std::max(a.values[i], b.values[i]);
+    c.set_value(i, std::min(a[i], b[i]));
```

Comment:
Didn't we decide to drop `[]` and `set_value`?

Reply:
I've made values private to make sure the packed types aren't treated like a vector. These functions provide default implementations when AVX/AVX2 is not around. Writing accessors makes it possible to delete them for template specializations to prevent runtime errors. This way we can know when something is trying to index a packed type. I'll remove set_value and replace it with a non-const version of operator[] since it's only meant to be used for the default implementation. This also makes for cleaner code.